### PR TITLE
Add check on field names

### LINKED ISSUE #13: Check illumination file contents
> Error checking where the contents of illfile are loaded, in
> particular, i don't currently check that the loaded file has the
> correct element names, only the number of elements

### PR DESCRIPTION
**Pull request by t-young31**
Closes 
- #13
- #14

### CODE REVIEW

**Comment by t-young31** [comment]
File used to test the function (based on examples/arc_01)

```matlab
addpath('../../tdms/matlab');
[x,y,z,lambda] = fdtd_bounds('pstd_input_file.m');

rad = 15e-6;
%refractive index of cylinder
refind = 1.42;
y = 0;
[X,Y,Z] = ndgrid(x,y,z);

I = zeros(size(X));
I( find( (X.^2 + Z.^2) < rad^2 )) = 1;
I( (end-3):end,1,:) = 0;
I( :, 1, (end-3):end) = 0;
inds = find(I(:));
[ii,jj,kk] = ind2sub(size(I), inds);
composition_matrix = [ii jj kk ones(size(ii))];
material_matrix = [1 refind^2 1 0 0 0     0     0     0 0 0];
save(sprintf('gridfiles/gridfile_cyl'), 'composition_matrix', 'material_matrix');

Isource = zeros(size(X));
Jsource = zeros(size(X));
Ksource = zeros(size(X));

save(sprintf('tmp'), 'Isource', 'Jsource', 'Jsource');

[fdtdgrid, Exs,Eys,Ezs,Hxs,Hys,Hzs,grid_labels,camplitudes,vertices,facets,Id] = ...
iteratefdtd_matrix('pstd_input_file.m','filesetup','in/pstd_cyl','gridfiles/gridfile_cyl.mat','tmp.mat');
```

**Review by giordano — APPROVED** [review]
Looks good to me, as far as I can tell (not super familiar with matlab though!).  My understanding is that the new check throws an error if the condition is not met, while previously the code will continue running, is the new behaviour intended?

**Comment by t-young31** [comment]
> is the new behaviour intended?

It's the behaviour I intended to write.. @prmunro is perhaps better placed to answer whether this is the intended outcome. Silently ignoring something is worse than throwing an error, in my opinion

**Comment by prmunro** [comment]
The above test won't result in the modified code being executed. I am in the process of generating a test that will.

**Comment by t-young31** [comment]
> The above test won't result in the modified code being executed. I am in the process of generating a test that will.

Indeed the previous script didn't check the exi/eyi could be present in the file. I've now added some specific unit tests which are run on Github actions. [Results from the last run](https://github.com/UCL/TDMS/runs/7558560046?check_suite_focus=true#step:4:9)

File: .github/workflows/matlab_tests.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  release:
+  workflow_dispatch:
+
+jobs:
+  tests:
+    name: Run MATLAB Tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v2
```
Comment (dstansby) [inline .github/workflows/matlab_tests.yml:16]:
```suggestion
        uses: actions/checkout@v3
```

File: tdms/tests/matlab/README.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+### MATLAB tests
+
+The unit tests in this directory can be run with
+
+```bash
+matlab -nodisplay -nosplash -nodesktop -r "run('run_tests.m');exit;"
```
Comment (dstansby) [inline tdms/tests/matlab/README.md:6]:
```suggestion
matlab -batch run_tests
```
This is how `MATLAB` recommends running a script without launching a window. I can't find the specific documentation right now I'm afraid (MATLAB documentaiton is a maze...)

File: tdms/tests/matlab/data/pstd_input_file.m [hunk 3]
```diff
@@ -0,0 +1,141 @@
+%these are not involved in the formal input file spec
```
Comment (dstansby) [inline tdms/tests/matlab/data/pstd_input_file.m:1]: Add a quick comment at the top of this file that explains what it is? Something like "Sample input file used for running unit tests" or similar I imagine.

File: tdms/tests/matlab/run_tests.m [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import matlab.unittest.selectors.HasBaseFolder;
+
+addpath(genpath('.'));
+
+suite = testsuite(pwd, 'IncludeSubfolders', true);
+suite = suite.selectIf(HasBaseFolder(ContainsSubstring('tests')));
+
+runner = TestRunner.withTextOutput('OutputDetail', Verbosity.Detailed);
+results = runner.run(suite);
+
```
Comment (dstansby) [inline tdms/tests/matlab/run_tests.m:13]: If you wnat to add code coverage (here or another PR):

https://github.com/dstansby/matlab-sandbox/blob/6da492672f240df162810cf9948b4a2bc8377841/run_tests.m#L22

File: tdms/tests/matlab/test_iteratefdtd_matrix.m [hunk 5]
```diff
@@ -0,0 +1,55 @@
+
+function tests = test_iteratefdtd_matrix_function
+tests = functiontests(localfunctions);
```
Comment (dstansby) [inline tdms/tests/matlab/test_iteratefdtd_matrix.m:3]:
```suggestion
    tests = functiontests(localfunctions);
```

File: tdms/tests/matlab/test_iteratefdtd_matrix.m [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+end
+
+function testInvalidIlluminationSourceIJKErrors(testCase)
+    verifyError(testCase, @()createInputWithInvalidIlluminationSourceIJ(), 'TDMSException:InvalidIlluminationFile');
+    delete *.mat
```
Comment (dstansby) [inline tdms/tests/matlab/test_iteratefdtd_matrix.m:8]: Probably better to create and then delete a specific directory?

File: tdms/tests/matlab/test_iteratefdtd_matrix.m [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    addPathAndCreateGridFile();
+    exi = zeros(size(1));
+    % Illumnation file must both exi and eyi
+    save(sprintf('invalid_illumnation_file'), 'exi', 'exi');
+
+    [~] = iteratefdtd_matrix('pstd_input_file.m','filesetup',...
+    'tmp_input','gridfile_cyl.mat','invalid_illumnation_file.mat');
+end
+
+function addPathAndCreateGridFile()
```
Comment (dstansby) [inline tdms/tests/matlab/test_iteratefdtd_matrix.m:39]: I think this deserves a quick comment explaining what it does at the top

**Review by dstansby — COMMENTED** [review]
Overall looks 👍 I've added some suggetions - enough for me to hold off approving for now.

File: tdms/tests/matlab/README.md (hunk 2 above)
Comment (t-young31) [reply tdms/tests/matlab/README.md:6]: love it 👍🏼

File: tdms/tests/matlab/run_tests.m (hunk 4 above)
Comment (t-young31) [reply tdms/tests/matlab/run_tests.m:13]: At this state I'm not too worries about adding codecov, but would definitely be nice in the future

File: tdms/tests/matlab/test_iteratefdtd_matrix.m (hunk 6 above)
Comment (t-young31) [reply tdms/tests/matlab/test_iteratefdtd_matrix.m:8]: Ah yes – the right way. Done now 🎉

**Review by dstansby — APPROVED** [review]

**Comment by prmunro** [comment]
[arc_04.zip](https://github.com/UCL/TDMS/files/9212413/arc_04.zip)

Here is a test (using top level file run_pstd_bscan.m) that shows when an error should and should not be generated. I agree that an error should be generated if an incorrect Ksource, Jsource or Isource.

**Comment by prmunro** [comment]
[arc_05.zip](https://github.com/UCL/TDMS/files/9219113/arc_05.zip)
Here is another test that shows correct usage of the data stucture with members exi and eyi, and one case where usage is incorrect and should generate an error.

**Comment by t-young31** [comment]
> Here is a test (using top level file run_pstd_bscan.m) that shows when an error should and should not be generated.

Many thanks for those – they've now been incorporated into the unit test suite

*** 
> Are these tests useful, or are they an overkill?

More tests are always useful (in my opinion at least). I'll add those in and re-request your review when it's done – thanks

**Review by prmunro — APPROVED** [review]

**Pull request merged**